pkg/cloudprovider/vsphereparavirtual/controllers/routablepod/ippool/ippool_controller.go

```diff
@@ -0,0 +1,297 @@
+	ippoolscheme "k8s.io/cloud-provider-vsphere/pkg/cloudprovider/vsphereparavirtual/client/clientset/versioned/scheme"
+	ippoolinformers "k8s.io/cloud-provider-vsphere/pkg/cloudprovider/vsphereparavirtual/client/informers/externalversions/nsxnetworking/v1alpha1"
+	ippoollisters "k8s.io/cloud-provider-vsphere/pkg/cloudprovider/vsphereparavirtual/client/listers/nsxnetworking/v1alpha1"
+	klog "k8s.io/klog/v2"
+)
+
+const (
+	controllerName    = "ippool-controller"
+	cidrUpdateRetries = 3
+	// Interval of synchronizing ippool status from apiserver
+	ippoolSyncPeriod = 30 * time.Second
+)
+
+type IPPoolController struct {
+	kubeclientset      kubernetes.Interface
+	ippoolclientset    ippoolclientset.Interface
+	ippoolLister       ippoollisters.IPPoolLister
+	ippoolListerSynced cache.InformerSynced
+
+	recorder  record.EventRecorder
+	workqueue workqueue.RateLimitingInterface
+}
+
+func NewIPPoolController(
+	kubeClient kubernetes.Interface,
+	ippoolclientset ippoolclientset.Interface,
+	ippoolInformer ippoolinformers.IPPoolInformer) *IPPoolController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(ippoolscheme.Scheme, corev1.EventSource{Component: controllerName})
+
+	ippc := &IPPoolController{
+		kubeclientset:      kubeClient,
+		ippoolclientset:    ippoolclientset,
+		ippoolLister:       ippoolInformer.Lister(),
+		ippoolListerSynced: ippoolInformer.Informer().HasSynced,
+
+		recorder:  recorder,
+		workqueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "IPPools"),
+	}
+
+	// watch ippool change
+	ippoolInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: ippc.enqueueIPPool,
+			UpdateFunc: func(old, cur interface{}) {
+				oldIppc, ok := old.(*ippoolv1alpha1.IPPool)
+				if !ok {
+					return
+				}
+				curIppc, ok := cur.(*ippoolv1alpha1.IPPool)
+				if !ok {
+					return
+				}
+				if !shouldSyncIPPool(oldIppc, curIppc) {
+					return
+				}
+				ippc.enqueueIPPool(cur)
+			},
+			// skip delete since network provider operator will clean up subnets
+		},
+		ippoolSyncPeriod,
+	)
+
+	return ippc
+}
+
+// if allocated subnets are updated, then need to update nodes with new subnets
+func shouldSyncIPPool(old, cur *ippoolv1alpha1.IPPool) bool {
+	return !reflect.DeepEqual(old.Status.Subnets, cur.Status.Subnets)
+}
+
+func (ippc *IPPoolController) enqueueIPPool(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	ippc.workqueue.Add(key)
+}
+
+func (ippc *IPPoolController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer ippc.workqueue.ShutDown()
+
+	klog.Infoln("Waiting cache to be synced.")
```

Comment:
`klog.V(4).Infoln` ?
we should probably determine the log level for each log 🤔

Reply:
yeah let me fix it